pkg/process/process.go

```diff
@@ -2,206 +2,191 @@ package process
+	if err != nil {
+		return err
+	}
 
-	exit := p.addExitWaiter()
-	select {
-	case <-ctx.Done():
-		err = ctx.Err()
-		break
-	case <-exit:
-	case <-f.Done():
-		break
+	e, err := res.Error()
+	if err != nil {
+		return err
 	}
 
-	return err
-}
+	switch e.Which() {
+	case api.Error_Which_none:
+		return nil
 
-// run the process and write to p.runDone after it is done.
-func (p *Process) run(ctx context.Context) {
-	var err error
-
-	// send the signal after the process finishes running
-	defer func() {
-		select {
-		case p.runDone <- err:
-			break
-		case <-ctx.Done():
-			p.runCancel()
+	case api.Error_Which_exitErr:
+		mod, err := e.ExitErr().Module()
+		if err != nil {
+			return err
 		}
-	}()
-
-	defer p.io.closeWriters(ctx)
+		return sys.NewExitError(mod, e.ExitErr().Code())
+	}
 
-	_, err = p.function.Call(p.runContext)
+	return fmt.Errorf("unknown error type: %d", e.Which())
 }
 
-// release calls all pending release functions.
-func (p *Process) release() {
-	for _, releaseFunc := range p.releaseFuncs {
-		defer releaseFunc()
-	}
+// process is the main implementation of the Process capability.
+type process struct {
+	fn     wasm.Function
+	handle procHandle
 }
 
 // Stop calls the runtime cancellation function.
-func (p *Process) Stop(ctx context.Context, call api.Process_stop) error {
-	p.runCancel()
-	return nil
+func (p *process) Stop(context.Context, api.Process_stop) error {
+	state := p.handle.Load()
+	if state.Err == nil {
+		state.Cancel()
+	}
+
+	return state.Err
 }
 
 // Start the process in the background.
-func (p *Process) Start(ctx context.Context, call api.Process_start) error {
-	go p.run(ctx)
+func (p *process) Start(_ context.Context, call api.Process_start) error {
+	state := p.handle.Load()
+	if state.Err != ErrNotStarted {
+		return state.Err
+	}
+
+	p.handle.Exec(p.fn)
 	return nil
 }
 
 // Wait for the process to finish running.
-func (p *Process) Wait(ctx context.Context, call api.Process_wait) error {
+func (p *process) Wait(ctx context.Context, call api.Process_wait) error {
+	state := p.handle.Load()
+	if state.Err == ErrNotStarted {
+		return state.Err
+	}
+
 	results, err := call.AllocResults()
 	if err != nil {
-		err = p.wait(ctx)
-		if err == nil {
-			results.SetError(proc_errors.Nil.Error())
-		} else {
-			results.SetError(err.Error())
-		}
+		return err
 	}
-	return err
-}
 
-// wait for the process to finish running.
-func (p *Process) wait(ctx context.Context) error {
-	var err error
+	call.Go()
+
 	select {
-	case err = <-p.runDone:
-		break
 	case <-ctx.Done():
-		break
+		return ctx.Err()
+	case <-state.Ctx.Done():
+		return p.handle.Bind(results)
 	}
-	return err
 }
 
-// moduleId retuns a shortened md5hash of the module
-func moduleId(binary []byte) string {
-	hash := md5.Sum(binary)
-	return hex.EncodeToString(hash[:])[:6]
+// procHandle encapsulates all the runtime state of a process.  Its
+// methods are safe for concurrent access.
+type procHandle atomic.Pointer[state]
+
+var (
```

Comment:
I believe it'll make the code more maintainable if we declare these kinds of variables that are treated as constants either at the beginning of the file or in a separate file :)

Reply:
Totally.  Let's move it to the top of the file, since there are only two of them.

Can I let you do it?